# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1225436086 -3600
#      Fri Oct 31 07:54:46 2008 +0100
# Branch release-3-0-x
# Node ID 7b69c809e0d426abe658b4029dc8910e7a3ca114
# Parent  57c9b3f2b82212325c5c81a9303a28aea5b15da2
oct-map.cc: copy key_list in indexing functions

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,12 @@
+2008-10-30  John W. Eaton  <jwe@octave.org>
+
+	* oct-map.cc (Octave_map::index): Copy key_list.
+
 2008-10-17  John W. Eaton  <jwe@octave.org>
 
 	* DLD-FUNCTIONS/__qp__.cc (qp): Undo part of change from
 	2007-09-04 (compute Y at each iteration).
 
 2008-10-02  John W. Eaton  <jwe@octave.org>
 
 	* pt-arg-list.cc (tree_argument_list::convert_to_const_vector):
diff --git a/src/oct-map.cc b/src/oct-map.cc
--- a/src/oct-map.cc
+++ b/src/oct-map.cc
@@ -482,16 +482,19 @@ Octave_map::index (const octave_value_li
 	{
 	  Cell tmp = contents(p).index (idx, resize_ok);
 
 	  if (error_state)
 	    break;
 
 	  retval.assign (key(p), tmp);
 	}
+
+      // Preserve order of keys.
+      retval.key_list = key_list;
     }
   else
     retval = *this;
 
   return retval;
 }
 
 Octave_map
@@ -504,16 +507,19 @@ Octave_map::index (idx_vector& i, int re
       Cell tmp = contents (p).index (i, resize_ok, rfv);
 
       if (error_state)
 	break;
 
       retval.assign (key (p), tmp);
     }
 
+  // Preserve order of keys.
+  retval.key_list = key_list;
+
   return retval;
 }
 
 Octave_map
 Octave_map::index (idx_vector& i, idx_vector& j, int resize_ok,
 		   const octave_value& rfv) const
 {
   Octave_map retval (dims ());
@@ -523,16 +529,19 @@ Octave_map::index (idx_vector& i, idx_ve
       Cell tmp = contents (p).index (i, j, resize_ok, rfv);
 
       if (error_state)
 	break;
 
       retval.assign (key (p), tmp);
     }
 
+  // Preserve order of keys.
+  retval.key_list = key_list;
+
   return retval;
 }
 
 Octave_map
 Octave_map::index (Array<idx_vector>& ra_idx, int resize_ok,
 		   const octave_value& rfv) const
 {
   Octave_map retval (dims ());
@@ -542,16 +551,19 @@ Octave_map::index (Array<idx_vector>& ra
       Cell tmp = contents (p).index (ra_idx, resize_ok, rfv);
 
       if (error_state)
 	break;
 
       retval.assign (key (p), tmp);
     }
 
+  // Preserve order of keys.
+  retval.key_list = key_list;
+
   return retval;
 }
 
 /*
 ;;; Local Variables: ***
 ;;; mode: C++ ***
 ;;; End: ***
 */
